server/src/main/java/org/elasticsearch/search/fetch/subphase/LookupField.java

```diff
@@ -0,0 +1,131 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Holds a specification and resolved values (if any) of {@link org.elasticsearch.index.mapper.LookupRuntimeFieldType} for a search hit.
+ * Data nodes create unresolved {@link LookupField} for each search hit during the fetch phase and coordinating nodes resolve them using
+ * {@link FetchLookupFieldsPhase}.
+ */
+public final class LookupField implements Writeable {
+    // Lookup request
+    private final String lookupIndex;
+    private final String lookupId;
+    private final List<String> lookupFields;
+
+    // Resolved values
+    private List<DocumentField> values;
+    private Exception failure;
+
+    LookupField(String lookupIndex, String lookupId, List<String> lookupFields) {
+        this.lookupIndex = Objects.requireNonNull(lookupIndex);
+        this.lookupId = Objects.requireNonNull(lookupId);
+        this.lookupFields = Objects.requireNonNull(lookupFields);
+    }
+
+    public LookupField(StreamInput in) throws IOException {
+        this.lookupIndex = in.readString();
+        this.lookupId = in.readString();
+        this.lookupFields = in.readStringList();
+        if (in.readBoolean()) {
+            this.values = in.readList(DocumentField::new);
+        }
+        this.failure = in.readException();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(lookupIndex);
+        out.writeString(lookupId);
+        out.writeStringCollection(lookupFields);
+        if (values != null) {
+            out.writeBoolean(true);
+            out.writeList(values);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeException(failure);
+    }
+
+    public boolean isResolved() {
+        return values != null || failure != null;
+    }
+
+    void setValues(List<DocumentField> values) {
+        assert isResolved() == false : "already resolved";
+        this.values = Objects.requireNonNull(values);
+    }
+
+    void setFailure(Exception failure) {
+        assert isResolved() == false : "already resolved";
+        this.failure = Objects.requireNonNull(failure);
+    }
+
+    List<DocumentField> getValues() {
+        return values;
+    }
+
+    Exception getFailure() {
+        return failure;
+    }
+
+    String getLookupIndex() {
+        return lookupIndex;
+    }
+
+    String getLookupId() {
+        return lookupId;
+    }
+
+    List<String> getLookupFields() {
+        return lookupFields;
+    }
+
+    /**
+     * Convert the resolved value or failure to a single value of a {@link DocumentField}
+     */
+    public Object asDocumentFieldValue() {
+        assert isResolved() : "unresolved lookup field";
+        final Map<String, Object> docFields = new HashMap<>();
+        docFields.put("_id", List.of(lookupId));
+        if (failure != null) {
+            docFields.put(
+                "_failure",
```

Comment:
I'm unsure about this part of the API where we handle failures. We don't have precedent for it in other parts of the search request, to return failures on a per-document or per-hit basis. It also seems like it could lead to the same failure message being returned several times -- for example if expensive queries were disabled, I think each hit would have the same error message (and the page size could be 100, so that would be 100 copies)?

We've brainstormed ideas for handling fine-grained errors in the context of runtime fields. Maybe for now we could have a really simple approach that matches the current error handling strategy. For most errors we'd error out the whole search. If the lookup document with the right ID was not found, this wouldn't cause the whole search to fail, but it would just be missing from the `fields` output.

Reply:
I am okay with your suggestion (maybe we can add an option `on_error` for this). Also, if we decide not to return the error, then I think we shouldn't return `_id`.